ngrok.go

```diff
@@ -43,20 +76,84 @@ func (n *Ngrok) Provision(ctx caddy.Context) error {
 	n.l = ctx.Logger()
 
 	if n.TunnelRaw == nil {
-		n.TunnelRaw = json.RawMessage(`{"tunnel": "tcp"}`)
+		n.TunnelRaw = json.RawMessage(`{"type": "tcp"}`)
 	}
+
 	tmod, err := ctx.LoadModule(n, "TunnelRaw")
 	if err != nil {
 		return fmt.Errorf("loading ngrok tunnel module: %v", err)
 	}
-	n.tunnel = tmod.(Tunnel)
+
+	var ok bool
+	n.tunnel, ok = tmod.(Tunnel)
+
+	if !ok {
+		return fmt.Errorf("loading ngrok tunnel module: %v", err)
+	}
+
+	err = n.doReplace()
+	if err != nil {
+		return fmt.Errorf("loading doing replacements: %v", err)
+	}
+
+	return nil
+}
+
+func (n *Ngrok) provisionOpts() error {
+	n.opts = append(n.opts, ngrok.WithLogger(ngrokZap.NewLogger(n.l)))
+
+	if n.AuthToken == "" {
+		n.opts = append(n.opts, ngrok.WithAuthtokenFromEnv())
+	} else {
+		n.opts = append(n.opts, ngrok.WithAuthtoken(n.AuthToken))
+	}
+
+	if n.Metadata != "" {
+		n.opts = append(n.opts, ngrok.WithMetadata(n.Metadata))
+	}
+
+	if n.Region != "" {
+		n.opts = append(n.opts, ngrok.WithRegion(n.Region))
+	}
+
+	if n.Server != "" {
+		n.opts = append(n.opts, ngrok.WithServer(n.Server))
+	}
+
+	n.opts = append(n.opts, ngrok.WithHeartbeatInterval(time.Duration(n.HeartbeatInterval)))
+
+	n.opts = append(n.opts, ngrok.WithHeartbeatTolerance(time.Duration(n.HeartbeatTolerance)))
+
+	return nil
+}
+
+func (n *Ngrok) doReplace() error {
+	repl := caddy.NewReplacer()
+	replaceableFields := []*string{
+		&n.AuthToken,
+		&n.Metadata,
+		&n.Region,
+		&n.Server,
+	}
+
+	for _, field := range replaceableFields {
+		actual, err := repl.ReplaceOrErr(*field, false, true)
+		if err != nil {
+			return fmt.Errorf("error replacing fields: %v", err)
+		}
+
+		*field = actual
+	}
+
+	return nil
+}
+
+// Validate implements caddy.Validator.
+func (n *Ngrok) Validate() error {
 	if n.tunnel == nil {
 		return fmt.Errorf("tunnel is required")
 	}
 
-	if repl, ok := ctx.Value(caddy.ReplacerCtxKey).(*caddy.Replacer); ok {
-		n.AuthToken = repl.ReplaceKnown(n.AuthToken, "")
-	}
 	return nil
 }
```

Comment:
I missed this in my earlier review. You're already checking if `n.tunnel` is proper and not nil in `Provision`, so this check will never happen. It's okay to remove this method entirely, including the interface guard at the bottom of the file.